File: study/ts/script/scene/view/siege/PopupSiegeRank.ts
const { ccclass, property } = cc._decorator;

import CommonRankIcon from '../../../ui/component/CommonRankIcon'

import CommonResourceInfo from '../../../ui/component/CommonResourceInfo'

import CommonTabGroupSmallHorizon from '../../../ui/component/CommonTabGroupSmallHorizon'

import CommonNormalMidPop from '../../../ui/component/CommonNormalMidPop'
import PopupBase from '../../../ui/PopupBase';
import { G_UserData, G_SignalManager, G_SceneManager } from '../../../init';
import { Lang } from '../../../lang/Lang';
import { handler } from '../../../utils/handler';
import { SignalConst } from '../../../const/SignalConst';
import PopupSiegeRankCell from './PopupSiegeRankCell';
import ListView from '../recovery/ListView';

@ccclass
export default class PopupSiegeRank extends PopupBase {

    @property({
        type: cc.Node,
        visible: true
    })
    _panelBase: cc.Node = null;

    @property({
        type: CommonNormalMidPop,
        visible: true
    })
    _rankBG: CommonNormalMidPop = null;

    @property({
        type: CommonTabGroupSmallHorizon,
        visible: true
    })
    _commonTabGroupSmallHorizon: CommonTabGroupSmallHorizon = null;

    @property({
        type: ListView,
        visible: true
    })
    _listRank: ListView = null;

    @property({
        type: cc.Label,
        visible: true
    })
    _textMyRank: cc.Label = null;

    @property({
        type: CommonResourceInfo,
        visible: true
    })
    _rankReward: CommonResourceInfo = null;

    @property({
        type: cc.Label,
        visible: true
    })
    _rewardTitle: cc.Label = null;

    @property({
        type: CommonRankIcon,
        visible: true
    })
    _myRankIcon: CommonRankIcon = null;

    @property({
        type: cc.Prefab,
        visible: true
    })
    _popupSiegeRankCellPrefab: cc.Prefab = null;

    private _data;
    private _personReward;
    private _guildReward;
    private _selectTabIndex;
    private _signalRank;
    private _signalGuildRank;
    private _rankData;
    private _type;

    public init(personReward, guildReward) {
        this._data = G_UserData.getSiegeRankData();
        this._personReward = personReward;
        this._guildReward = guildReward;
        this._selectTabIndex = 0;
        this._signalRank = null;
        this._signalGuildRank = null;
        this._rankData = null;
        this._type = null;
        this.setClickOtherClose(true);
    }

    public onCreate() {
        this._rankBG.setTitle(Lang.get('siege_rank_title'));
        this._rankBG.addCloseEventListener(handler(this, this._onCloseClick));
        this._initTab();
    }

    public onEnter() {
        this._refreshMyRank();
        this._signalRank = G_SignalManager.add(SignalConst.EVENT_SIEGE_RANK, handler(this, this._onEventSiegeRank));
        this._signalGuildRank = G_SignalManager.add(SignalConst.EVENT_SIEGE_GUILD_RANK, handler(this, this._onEventSiegeGuildRank));
        G_UserData.getSiegeRankData().c2sGetRebelArmyHurtRank();
        var runningScene = G_SceneManager.getRunningScene();
        runningScene.addGetUserBaseInfoEvent();
    }

    public onExit() {
        this._signalRank.remove();
        this._signalRank = null;
        this._signalGuildRank.remove();
        this._signalGuildRank = null;
    }

    private _onCloseClick() {
        this.closeWithAction();
    }

    private _refreshMyRank() {
        var myRank = G_UserData.getSiegeData().getUserRank();
        if (myRank == 0) {
            this._textMyRank.string = (Lang.get('siege_rank_no_rank'));
            this._textMyRank.node.active = true;
            this._myRankIcon.node.active = false;
        } else if (myRank < 4) {
            this._textMyRank.node.active = false;
            this._myRankIcon.node.active = true;
            this._myRankIcon.setRank(myRank);
        } else {
            this._textMyRank.string = (myRank).toString();
            this._textMyRank.node.active = true;
            this._myRankIcon.node.active = false;
        }
        if (myRank != 0) {
            var reward = this._personReward;
            if (reward) {
                this._rankReward.updateUI(reward.type, reward.value, reward.size);
                this._rankReward.node.active = true;
                this._rewardTitle.node.active = true;
            }
        } else {
            this._rewardTitle.node.active = false;
            this._rankReward.node.active = false;
        }
    }

    private _refreshMyGuildRank() {
        var myGuildRank = G_UserData.getSiegeData().getUserGuildRank();
        if (myGuildRank == 0) {
            this._textMyRank.string = (Lang.get('siege_rank_no_rank'));
            this._textMyRank.node.active = true;
            this._myRankIcon.node.active = false;
        } else if (myGuildRank < 4) {
            this._myRankIcon.setRank(myGuildRank);
            this._textMyRank.node.active = false;
            this._myRankIcon.node.active = true;
        } else {
            this._textMyRank.string = (myGuildRank).toString();
            this._textMyRank.node.active = true;
            this._myRankIcon.node.active = false;
        }
        if (myGuildRank != 0) {
            var reward = this._guildReward;
            if (reward) {
                this._rankReward.updateUI(reward.type, reward.value, reward.size);
                this._rankReward.node.active = true;
                this._rewardTitle.node.active = true;
            }
        } else {
            this._rewardTitle.node.active = false;
            this._rankReward.node.active = false;
        }
    }

    private _refreshPersonRank() {
        this._listRank.clearAll();
        this._rankData = this._data.getRankDatas();
        this._type = PopupSiegeRankCell.TYPE_PERSON;
        var listView = this._listRank;
        listView.setTemplate(this._popupSiegeRankCellPrefab);
        listView.setCallback(handler(this, this._onRankUpdate));
        listView.setData(this._rankData);
    }

    private _refreshGuildRank() {
        this._listRank.clearAll();
        this._rankData = this._data.getGuildRankDatas();
        this._type = PopupSiegeRankCell.TYPE_GUILD;
        var listView = this._listRank;
        listView.setTemplate(this._popupSiegeRankCellPrefab);
        listView.setCallback(handler(this, this._onRankUpdate));
        listView.setData(this._rankData);
    }

    private _onRankUpdate(node:cc.Node, index) {
        let item:PopupSiegeRankCell = node.getComponent(PopupSiegeRankCell);
        var rankData = this._rankData;
        if (rankData.length > 0) {
            item.refreshInfo(rankData[index], this._type);
        }
    }

    private _onItemSelected(item, index) {
    }

    private _initTab() {
        var param = {
            callback: handler(this, this._onTabSelect),
            isVertical: 2,
            offset: 2,
            textList: [
                Lang.get('siege_rank_type1'),
                Lang.get('siege_rank_type2')
            ]
        };
        this._commonTabGroupSmallHorizon.recreateTabs(param);
        this._refreshPersonRank();
    }

    private _onTabSelect(index) {
        if (this._selectTabIndex == index) {
            return;
        }
        this._selectTabIndex = index;
        if (index == 0) {
            G_UserData.getSiegeRankData().c2sGetRebelArmyHurtRank();
            this._refreshMyRank();
        } else if (index == 1) {
            G_UserData.getSiegeRankData().c2sGetRebelArmyGuildHurtRank();
            this._refreshMyGuildRank();
        }
    }

    private _onEventSiegeRank() {
        this._refreshPersonRank();
    }

    private _onEventSiegeGuildRank() {
        this._refreshGuildRank();
    }
}

File: study/lua/src/app/scene/view/siege/PopupSiegeRank.lua
local PopupBase = require("app.ui.PopupBase")
local PopupSiegeRank = class("PopupSiegeRank", PopupBase)

local PopupSiegeRankCell = require("app.scene.view.siege.PopupSiegeRankCell")
local Hero = require("app.config.hero")
local Color = require("app.utils.Color")
local TabButtonGroup = require("app.utils.TabButtonGroup")
local RebelRankReward = require("app.config.rebel_rank_reward")

function PopupSiegeRank:ctor(personReward, guildReward)
    self._data = G_UserData:getSiegeRankData()	--排行数据
	self._personReward = personReward
	self._guildReward = guildReward

    self._rankBG = nil  --背景
	self._commonTabGroupSmallHorizon = nil

	self._textMyRank = nil		--我的排名
	self._rewardTitle = nil 	--排名奖励
	self._rankReward = nil		--奖励内容
	self._myRankIcon = nil		--排行icon
	self._listRank = nil		--排行list
	self._selectTabIndex = 1 	--初始切页

	--signal
	self._signalRank = nil			--叛军排行
	self._signalGuildRank = nil		--叛军工会排行

	self._rankData = nil
	self._type = nil 			

	local resource = {
		file = Path.getCSB("PopupSiegeRank", "siege"),
		binding = {
		}
	}
	PopupSiegeRank.super.ctor(self, resource)
end

function PopupSiegeRank:onCreate()
	self._rankBG:setTitle(Lang.get("siege_rank_title"))
	self._rankBG:addCloseEventListener(handler(self, self._onCloseClick))
	self:_initTab()
end

function PopupSiegeRank:onEnter()
	self:_refreshMyRank()
	self._signalRank = G_SignalManager:add(SignalConst.EVENT_SIEGE_RANK, handler(self, self._onEventSiegeRank))
	self._signalGuildRank = G_SignalManager:add(SignalConst.EVENT_SIEGE_GUILD_RANK, handler(self, self._onEventSiegeGuildRank))

	G_UserData:getSiegeRankData():c2sGetRebelArmyHurtRank()	--初始化的时候，显示个人排行
	local runningScene = G_SceneManager:getRunningScene()
	runningScene:addGetUserBaseInfoEvent()
end

function PopupSiegeRank:onExit()
	self._signalRank:remove()
	self._signalRank = nil
	self._signalGuildRank:remove()
	self._signalGuildRank = nil
end

function PopupSiegeRank:_onCloseClick()
    self:closeWithAction()
end

function PopupSiegeRank:_refreshMyRank()
	local myRank = G_UserData:getSiegeData():getUserRank()
	if myRank == 0 then
		self._textMyRank:setString(Lang.get("siege_rank_no_rank"))
		self._textMyRank:setVisible(true)
		self._myRankIcon:setVisible(false)
	elseif myRank < 4 then
		self._textMyRank:setVisible(false)
		self._myRankIcon:setVisible(true)
		self._myRankIcon:setRank(myRank)
	else
        self._textMyRank:setString(myRank)
		self._textMyRank:setVisible(true)
		self._myRankIcon:setVisible(false)
	end
	if myRank ~= 0 then
    	local reward = self._personReward
		if reward then
			self._rankReward:updateUI(reward.type,reward.value, reward.size) 
			self._rankReward:setVisible(true)
			self._rewardTitle:setVisible(true)
		end
	else
		self._rewardTitle:setVisible(false)
		self._rankReward:setVisible(false)
	end
end

function PopupSiegeRank:_refreshMyGuildRank()
	local myGuildRank = G_UserData:getSiegeData():getUserGuildRank()
	if myGuildRank == 0 then
		self._textMyRank:setString(Lang.get("siege_rank_no_rank"))
		self._textMyRank:setVisible(true)
        self._myRankIcon:setVisible(false)
	elseif myGuildRank < 4 then
        self._myRankIcon:setRank(myGuildRank)      
        self._textMyRank:setVisible(false)
        self._myRankIcon:setVisible(true)
	else
        self._textMyRank:setString(myGuildRank)
        self._textMyRank:setVisible(true)
        self._myRankIcon:setVisible(false)
	end
	if myGuildRank ~= 0 then
		local myRankString = tostring(myGuildRank)
		local reward = self._guildReward
		if reward then
			self._rankReward:updateUI(reward.type,reward.value, reward.size) 
			self._rankReward:setVisible(true)
			self._rewardTitle:setVisible(true)
		end
	else
		self._rewardTitle:setVisible(false)
		self._rankReward:setVisible(false)
	end
end

function PopupSiegeRank:_refreshPersonRank()
	self._listRank:clearAll()
	self._rankData = self._data:getRankDatas()
	self._type = PopupSiegeRankCell.TYPE_PERSON
	local listView = self._listRank
    listView:setTemplate(PopupSiegeRankCell)
    listView:setCallback(handler(self, self._onRankUpdate), handler(self, self._onItemSelected))
    listView:resize(#self._rankData)
end

function PopupSiegeRank:_refreshGuildRank()
	self._listRank:clearAll()
	self._rankData = self._data:getGuildRankDatas()
	self._type = PopupSiegeRankCell.TYPE_GUILD
	local listView = self._listRank
    listView:setTemplate(PopupSiegeRankCell)
    listView:setCallback(handler(self, self._onRankUpdate), handler(self, self._onItemSelected))
    listView:resize(#self._rankData)
end

function PopupSiegeRank:_onRankUpdate(item, index)
    local rankData = self._rankData
    if #rankData > 0 then
		item:refreshInfo(rankData[index+1], self._type)
	end
end

function PopupSiegeRank:_onItemSelected(item, index)

end

function PopupSiegeRank:_initTab()
	local param = {
		callback = handler(self, self._onTabSelect),
		isVertical = 2,
		offset = 2,
		textList = {Lang.get("siege_rank_type1"), Lang.get("siege_rank_type2")}
	}
	self._commonTabGroupSmallHorizon:recreateTabs(param)

	self:_refreshPersonRank()
end

function PopupSiegeRank:_onTabSelect(index)
	if self._selectTabIndex == index then
		return
	end
	self._selectTabIndex = index
	if index == 1 then			--个人排行
		G_UserData:getSiegeRankData():c2sGetRebelArmyHurtRank()
		self:_refreshMyRank()
	elseif index == 2 then		--军团排行
		G_UserData:getSiegeRankData():c2sGetRebelArmyGuildHurtRank()
		self:_refreshMyGuildRank()
	end
end

function PopupSiegeRank:_onEventSiegeRank()
	self:_refreshPersonRank()
end

function PopupSiegeRank:_onEventSiegeGuildRank()
	self:_refreshGuildRank()
end


return PopupSiegeRank